![My Image](https://raw.githubusercontent.com/ralf-42/Image/main/genai-banner-2.jpg)

<p><font size="5" color='grey'> <b>
Multimodales RAG
</b></font> </br></p>


---

In [1]:
#@title 🔧 Umgebung einrichten{ display-mode: "form" }
!uv pip install --system -q git+https://github.com/ralf-42/GenAI.git#subdirectory=04_modul
from genai_lib.utilities import check_environment, get_ipinfo, setup_api_keys, mprint, install_packages
setup_api_keys(['OPENAI_API_KEY', 'HF_TOKEN'], create_globals=False)
print()
check_environment()
print()
get_ipinfo()

✓ OPENAI_API_KEY erfolgreich gesetzt
✓ HF_TOKEN erfolgreich gesetzt

Python Version: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]

Installierte LangChain-Bibliotheken:
langchain                                1.0.3
langchain-chroma                         1.0.0
langchain-classic                        1.0.0
langchain-community                      0.4.1
langchain-core                           1.0.3
langchain-ollama                         1.0.0
langchain-openai                         1.0.2
langchain-text-splitters                 1.0.0

Installierte LangGraph-Bibliotheken:
langgraph                                1.0.2
langgraph-checkpoint                     3.0.0
langgraph-prebuilt                       1.0.2
langgraph-sdk                            0.2.9

IP-Adresse: 34.136.75.120
Hostname: 120.75.136.34.bc.googleusercontent.com
Stadt: Council Bluffs
Region: Iowa
Land: US
Koordinaten: 41.2619,-95.8608
Provider: AS396982 Google LLC
Postleitzahl: 51502
Zeitzone: America/Chicag

In [2]:
#@title 🛠️ Installationen { display-mode: "form" }
install_packages([
    ('markitdown[all]', 'markitdown'),
])

🔄 Installiere markitdown[all]...
✅ markitdown[all] erfolgreich installiert und importiert
✅ langchain_chroma bereits verfügbar


In [3]:
#@title 📂 Dokumente, Bilder { display-mode: "form" }
!rm -rf files
!mkdir files

# --- Texte
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/01_text/biografien_1.txt -o files/biografien_1.txt
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/01_text/biografien_2.md -o files/biografien_2.md
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/01_text/biografien_3.pdf -o files/biografien_3.pdf
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/01_text/biografien_4.docx -o files/biografien_4.docx
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/01_text/roboter.txt -o files/roboter.txt

# --- Bilder
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/02_bild/retro_robot.jpg -o files/retro_robot.jpg
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/02_bild/hedra_cyborg.png -o files/hedra_cyborg.png
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/02_bild/apfel.png -o files/apfel.png
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02_daten/02_bild/zwei_roboter.png -o zwei_roboter.png

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5324  100  5324    0     0  13218      0 --:--:-- --:--:-- --:--:-- 13243
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7868  100  7868    0     0  44440      0 --:--:-- --:--:-- --:--:-- 44704
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 68184  100 68184    0     0   224k      0 --:--:-- --:--:-- --:--:--  224k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 18743  100 18743    0     0   103k      0 --:--:-- --:--:-- --:--:--  103k
  % Total    % Received % Xferd  Average Speed   Tim

# 1 | RAG-Prozess
---

In dem entwickelten **multimodalen Retrieval-Augmented-Generation (RAG)**-System werden sowohl Text- als auch Bilddaten verarbeitet.
Für die **Bildverarbeitung** kommt ein zweistufiger Ansatz zum Einsatz:

1. **Image-Embedding:**
   Das Bild wird in einen hochdimensionalen Vektorraum eingebettet, um visuelle Merkmale wie Formen, Farben und Strukturen numerisch zu repräsentieren.

2. **Textbeschreibung und Text-Embedding:**
   Zusätzlich wird mit *gpt-4o-mini* eine sprachliche Beschreibung des Bildes erzeugt. Diese Beschreibung wird anschließend in ein Text-Embedding überführt, um semantische Informationen textbasiert nutzbar zu machen.

---

**Vorteile des Ansatzes:**

* **Erweiterte semantische Repräsentation:**
  Durch die Kombination von visuellen und sprachlichen Embeddings werden sowohl konkrete als auch konzeptuelle Eigenschaften eines Bildes abgebildet.

* **Verbesserte Retrieval-Qualität:**
  Textbasierte Suchanfragen können nicht nur über visuelle Ähnlichkeiten, sondern auch über die semantisch beschriebene Bedeutung der Bilder beantwortet werden.

* **Höhere Interpretierbarkeit:**
  Die generierte Bildbeschreibung ermöglicht eine transparente Nachvollziehbarkeit der zugrunde liegenden Repräsentationen und unterstützt bei der Evaluierung der Ergebnisse.

* **Gute Erweiterbarkeit:**
  Das Verfahren ist modular aufgebaut und lässt sich leicht um weitere Modalitäten wie Audio oder Video ergänzen.


<p><font color='black' size="5">
RAG-Prozess für Texte
</font></p>

<img src="https://raw.githubusercontent.com/ralf-42/Image/main/rag_process.png" width="600" alt="Avatar">

<p><font color='black' size="5">
RAG-Prozess für Bilder
</font></p>

<img src="https://raw.githubusercontent.com/ralf-42/Image/main/rag_process_03.png" width="600" alt="Avatar">

# 2 | Modul `multimodal_rag`
---

Python-Modul für ein **funktionales multimodales RAG-System**, das Text- und Bilddokumente in einer einheitlichen Vektordatenbank verwaltet und durchsucht.

**Modalitätsrichtungen**

| Eingabe (Query) | Ausgabe (Antwort) | Beispiel / Beschreibung | Status |
|-----------------|-------------------|-------------------------|--------|
| **Text → Text** | Textbasierte Frage führt zu Textantwort | Klassisches RAG-System (z.B. Chatbot, Q&A) | ✅ |
| **Text → Bild** | Textanfrage findet relevante Bilder | "Zeige mir Roboter-Bilder" | ✅ |
| **Bild → Text** | Bildanalyse oder Captioning | "Was ist auf diesem Foto zu sehen?" | ✅ |
| **Bild → Bild** | Bildretrieval oder visuelle Transformation | "Finde ähnliche Bilder" | ✅ |
| **Text + Bild → Text** | Kombination zur Textgenerierung | "Welche Informationen enthält dieses Diagramm?" | ❌ |
| **Text + Bild → Bild** | Bedingte Bildgenerierung | "Mach aus diesem Bild eine Winterversion" | ❌ |

**Hauptvorteile**

1. **Funktionale Architektur**: Klare Trennung von Konfiguration, Komponenten und Funktionen
2. **Einheitliche Datenbank**: ChromaDB mit separaten Collections für Text und Bilder
3. **Hybride Suche**: Text-Embeddings (OpenAI) + Bild-Embeddings (CLIP)
4. **Flexible Konfiguration**: Alle Parameter über RAGConfig anpassbar
5. **Bildbeschreibung**: Zu jedem Bild wird zusätzlich eine Bildbeschreibung erstellt.

<p><font color='black' size="5">
Aufbau
</font></p>

**RAGConfig**
- Zentrale Konfigurationsklasse
- Anpassbare Parameter (chunk_size, models, thresholds)

**RAGComponents**
- Container für alle System-Komponenten
- Text-Embeddings, CLIP-Model, LLMs, Collections

**Hauptfunktionen**
- Datensammlung
    - `init_rag_system()`: System-Initialisierung
    - `process_directory()`: Bulk-Import von Dateien
    - `add_text_document()`: Einzelnes Dokument hinzufügen
    - `add_image_with_description()`: Bild mit Auto-Beschreibung
- Abruf & Erweiterung
    - `search_texts()`: Text-Suche inkl. Bildbeschreibungen
    - `search_images()`: CLIP-basierte Bildsuche
    - `search_similar_images()`: Bild→Bild Ähnlichkeitssuche
    - `search_text_by_image()`: Bild→Text Suche
    - `multimodal_search()`: Erweiterte multimodale Suche

<img src="https://raw.githubusercontent.com/ralf-42/Image/main/multimodal_rag_function_call.png" width="1100" alt="Avatar">




In [4]:
# Import des Moduls
from genai_lib.multimodal_rag import (
    init_rag_system,
    process_directory,
    multimodal_search,
    search_texts,
    search_images,
    search_text_by_image,
    search_similar_images,
    add_text_document,
    add_image_with_description,
    get_system_status
)

# 3 | RAG initialisieren
---

In [5]:
# Initialisierung
rag = init_rag_system()

🚀 Initialisiere Enhanced RAG-System in ./multimodal_rag_db
✅ OpenAI Text-Embeddings initialisiert
🖼️ Lade CLIP-Modell...


modules.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

config.json: 0.00B [00:00, ?B/s]

tokenizer_config.json:   0%|          | 0.00/604 [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/316 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/389 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

0_CLIPModel/model.safetensors:   0%|          | 0.00/605M [00:00<?, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

0_CLIPModel/pytorch_model.bin:   0%|          | 0.00/605M [00:00<?, ?B/s]

Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.


✅ CLIP-Modell geladen
✅ LLMs initialisiert (Text + Vision)
✅ Collections initialisiert



In [6]:
# Konfiguration abfragen
print(rag.config)

# Konfiguration ändern (Beispiel: chunk_size ändern)
# rag.config.chunk_size = 300

RAGConfig(chunk_size=200, chunk_overlap=20, text_threshold=1.2, image_threshold=0.8, clip_model='clip-ViT-B-32', text_model='text-embedding-3-small', llm_model='gpt-4o-mini', vision_model='gpt-4o-mini', db_path='./multimodal_rag_db')


# 4 | Bilder/Texte in RAG übernehmen
---

In [7]:
# Dokumente/Bilder laden & verarbeiten
process_directory(rag, './files', auto_describe_images=True)

📊 Gefunden: 5 Text-Dateien, 3 Bilder

📄 biografien_1.txt
✅ 32 Chunks von 'biografien_1.txt' hinzugefügt
📄 biografien_3.pdf
✅ 68 Chunks von 'biografien_3.pdf' hinzugefügt
📄 roboter.txt
✅ 11 Chunks von 'roboter.txt' hinzugefügt
📄 biografien_4.docx
✅ 63 Chunks von 'biografien_4.docx' hinzugefügt
📄 biografien_2.md
✅ 44 Chunks von 'biografien_2.md' hinzugefügt

🖼️ apfel.png
🖼️ Bild-Embedding erstellt für apfel.png
📝 Bildbeschreibung generiert: Das Bild zeigt einen glänzenden roten Apfel, der auf einer Holzoberfläche platziert ist. Die Farben ...
✅ Bildbeschreibung in Text-Collection gespeichert
✅ Bild 'apfel.png' mit Cross-References hinzugefügt

🖼️ hedra_cyborg.png
🖼️ Bild-Embedding erstellt für hedra_cyborg.png
📝 Bildbeschreibung generiert: Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metall...
✅ Bildbeschreibung in Text-Collection gespeichert
✅ Bild 'hedra_cyborg.png' mit Cross-References hinzugefügt

🖼️ retro_robot.jpg
🖼️ Bild-Embedding e

{'texts': 5, 'images': 3, 'image_descriptions': 3}

# 5 | Suche mit Text
---

<p><font color='black' size="5">
Text → Text
</font></p>

In [8]:
result = search_texts(rag, "Was weisst Du über Cyborgs?")
mprint(result)

### 🤖 LLM-Antwort

Cyborgs sind hybride Wesen, die sowohl biologische als auch technologische Komponenten vereinen. Sie können Menschen sein, die durch mechanische oder elektronische Teile ergänzt wurden, um ihre Fähigkeiten zu erweitern oder zu verbessern. In der Science-Fiction werden Cyborgs oft als Charaktere dargestellt, die übermenschliche Kräfte oder Fähigkeiten besitzen, was sie in verschiedenen Erzählungen zu zentralen Figuren macht. Die ethischen und gesellschaftlichen Implikationen von Cyborgs sind vielfältig, insbesondere in Bezug auf Identität, Verantwortung und die Grenzen zwischen Mensch und Maschine.

### 📚 Verwendete Quellen

   📄 Text-Dokumente (2):
      1. roboter.txt (Ähnlichkeit: 0.444)
      2. roboter.txt (Ähnlichkeit: 0.43)

   🖼️ Bildbeschreibungen (1):
      1. hedra_cyborg.png (Ähnlichkeit: 0.545)


<p><font color='black' size="5">
Text → Bild
</font></p>

In [ ]:
# search_images() gibt jetzt formatierte String-Antwort mit LLM-Response zurück
result = search_images(rag, "Was weisst Du über Cyborgs?")
mprint(result)

<p><font color='black' size="5">
Text → Text/Bild
</font></p>

In [10]:
result = multimodal_search(rag, "Was weisst Du über Cyborgs?")
mprint(result)


🔍 Multimodale Suche: Was weisst Du über Cyborgs?



### 🤖 LLM-Antwort

Cyborgs sind hybride Wesen, die sowohl biologische als auch technologische Komponenten vereinen. Sie können Menschen sein, die durch mechanische oder elektronische Teile ergänzt wurden, um ihre Fähigkeiten zu erweitern oder zu verbessern. In der Science-Fiction werden Cyborgs oft als Charaktere dargestellt, die übermenschliche Kräfte oder Fähigkeiten besitzen, während sie gleichzeitig mit menschlichen Emotionen und Erfahrungen verbunden sind. Die ethischen und gesellschaftlichen Fragen, die sich im Zusammenhang mit Cyborgs stellen, betreffen unter anderem Identität, Verantwortung und die Auswirkungen der Technologie auf das menschliche Leben.

### 📚 Verwendete Quellen (via semantische Textsuche)

   📄 Text-Dokumente (2):
      1. roboter.txt (Ähnlichkeit: 0.444)
      2. roboter.txt (Ähnlichkeit: 0.43)

   🖼️ Bildbeschreibungen (1):
      1. hedra_cyborg.png (Ähnlichkeit: 0.545)

### 🔍 Bilder via CLIP (direkte visuelle Ähnlichkeitssuche)
   1. hedra_cyborg.png (Ähnlichkeit: 0.239)
      📝 Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metallische Elemente aufweist. Die dominierenden Farben sind dunkles Blau und Schwarz, was eine ernste und kraftvolle Stimmung vermittelt. Der Hintergrund ist unscharf, lässt jedoch auf eine moderne, lichtd...
   2. retro_robot.jpg (Ähnlichkeit: 0.22)
      📝 Das Bild zeigt einen freundlichen, silbernen Roboter, der vor einem Gebäude steht, das mit einem Schild "WELCOME ROBOTS WELCOME" geschmückt ist. Die Farben sind hell und einladend, mit warmen Tönen, die eine positive und futuristische Stimmung erzeugen. Die Komposition zeigt den Roboter im Vordergru...

### 🔗 Bilder via Cross-Modal Retrieval (Text → Bild Verknüpfung)
   1. hedra_cyborg.png
      📝 Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metallische Elemente aufweist. Die dominierenden Farben sind dunkles Blau und Schwarz, was eine ernste und kraftvolle Stimmung vermittelt. Der Hintergrund ist unscharf, lässt jedoch auf eine moderne, lichtd...


# 6 | Suche mit Bild
---

<p><font color='black' size="5">
Bild → Bild
</font></p>

In [11]:
# Bild → Bild: Finde visuell ähnliche Bilder
result = search_similar_images(rag, "./zwei_roboter.png", k=5)
mprint("## 🖼️ Suche Bild → Bild")
mprint("---")
for img in result:
    mprint(f"{img['filename']}: Ähnlichkeit: {img['similarity']}")

🔍 CLIP-Embedding erstellt für Query-Bild: zwei_roboter.png
✅ 3 ähnliche Bilder gefunden



## 🖼️ Suche Bild → Bild

---

retro_robot.jpg: Ähnlichkeit: 0.829

apfel.png: Ähnlichkeit: 0.677

hedra_cyborg.png: Ähnlichkeit: 0.469

<p><font color='black' size="5">
Bild → Text
</font></p>

In [12]:
# Bild → Text: Finde Textinformationen zu dem Bild
result = search_text_by_image(rag, "./zwei_roboter.png", k=5)
mprint(result)


🔍 Bild → Text Suche für: zwei_roboter.png

🔍 CLIP-Embedding erstellt für Query-Bild: zwei_roboter.png
✅ 3 ähnliche Bilder gefunden

🔍 Suche nach relevanten Text-Dokumenten basierend auf Bildbeschreibungen...
✅ 3 relevante Text-Dokumente gefunden



### 🤖 LLM-Antwort

Die Analyse des Query-Bildes zeigt visuelle und semantische Ähnlichkeiten zu mehreren Bildern und Textdokumenten, die sich alle um das Thema Roboter und deren Darstellung in verschiedenen Kontexten drehen.

**Visuelle Ähnlichkeiten:**
1. **retro_robot.jpg**: Dieses Bild zeigt einen freundlichen, silbernen Roboter in einer einladenden Umgebung, die auf eine positive Zukunftsvision hinweist. Die hellen Farben und die retro-futuristische Gestaltung des Gebäudes unterstreichen die Thematik der Robotik als Teil einer lebendigen, modernen Welt.
2. **hedra_cyborg.png**: Hier wird ein männlicher Charakter in einem futuristischen Anzug dargestellt, was auf eine ernstere, technologisch fortgeschrittene Perspektive auf Roboter und Cyborgs hinweist. Die dunklen Farben und der entschlossene Gesichtsausdruck vermitteln eine kraftvolle Stimmung, die oft in Science-Fiction-Kontexten zu finden ist.

**Gemeinsamkeiten:**
- Alle Bilder thematisieren Roboter und deren Integration in die Gesellschaft, sei es in einem freundlichen, einladenden Kontext oder in einem ernsten, technologischen Setting.
- Die Verwendung von Farben und Kompositionen in den Bildern schafft unterschiedliche Stimmungen, die von einladend und positiv bis hin zu ernst und futuristisch reichen.

**Relevante Kontextinformationen aus den Texten:**
- Die Textdokumente betonen, dass Roboter nicht nur technische Werkzeuge sind, sondern auch Ausdruck menschlicher Kreativität und Ambition. Sie reflektieren, wie wir unsere Zukunft und die Rolle der Technologie in unserem Leben sehen.
- Der Wandel im Bild des Roboters wird hervorgehoben, von klassischen Industrierobotern hin zu modernen Servicerobotern und autonomen Fahrzeugen, was die Vielfalt und die sich verändernde Wahrnehmung von Robotern in der Gesellschaft verdeutlicht.
- Roboter sind mittlerweile ein fester Bestandteil des Alltags und der industriellen Produktion, was die Relevanz und die Notwendigkeit ihrer Integration in verschiedene Lebensbereiche unterstreicht.

Insgesamt zeigen die Informationen, dass Roboter in verschiedenen Formen und Kontexten dargestellt werden, wobei sowohl positive als auch ernsthafte Perspektiven auf ihre Rolle in der Zukunft und im Alltag vermittelt werden.

### 🔍 Verwendete Quellen (visuell ähnliche Bilder via CLIP)
   1. retro_robot.jpg (Ähnlichkeit: 0.829)
      📝 Das Bild zeigt einen freundlichen, silbernen Roboter, der vor einem Gebäude steht, das mit einem Schild "WELCOME ROBOTS WELCOME" geschmückt ist. Die Farben sind hell und einladend, mit warmen Tönen, die eine positive und futuristische Stimmung erzeugen. Die Komposition zeigt den Roboter im Vordergru...
   2. apfel.png (Ähnlichkeit: 0.677)
      📝 Das Bild zeigt einen glänzenden roten Apfel, der auf einer Holzoberfläche platziert ist. Die Farben sind lebhaft, mit einem satten Rot des Apfels und einem sanften, neutralen Hintergrund, der eine warme und einladende Stimmung erzeugt. Die Komposition ist einfach und fokussiert, wobei der Apfel zent...
   3. hedra_cyborg.png (Ähnlichkeit: 0.469)
      📝 Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metallische Elemente aufweist. Die dominierenden Farben sind dunkles Blau und Schwarz, was eine ernste und kraftvolle Stimmung vermittelt. Der Hintergrund ist unscharf, lässt jedoch auf eine moderne, lichtd...

### 📚 Relevante Text-Dokumente (via semantische Suche)
   1. roboter.txt (Ähnlichkeit: 0.525)
      📄 Roboter sind damit nicht nur technische Werkzeuge, sondern Ausdruck menschlicher Kreativität und Ambition – sie spiegeln wider, wie wir uns selbst, unsere Arbeit und unsere Zukunft verstehen. Der...
   2. roboter.txt (Ähnlichkeit: 0.484)
      📄 kann. Gleichzeitig verändert sich das Bild des Roboters zunehmend: Neben den klassischen Industrierobotern treten heute Serviceroboter, autonome Fahrzeuge und kollaborative Roboter („Cobots“) auf,...
   3. roboter.txt (Ähnlichkeit: 0.427)
      📄 Roboter sind längst keine Zukunftsvision mehr, sondern fester Bestandteil unseres Alltags und der industriellen Produktion. Sie übernehmen Aufgaben, die für Menschen zu gefährlich, monoton oder...


# 7 | Aufgabe
---

<p><font color='black' size="5">
Teste das Multimodale RAG-System mit eigenen Queries
</font></p>

**Ziel**: Nutze das fertige multimodale RAG-System, um verschiedene Such-Operationen durchzuführen.



**Schritt 1: Einfache Queries testen**

Führe die folgenden Queries mit dem RAG-System aus und beobachte, welche Ergebnisse zurückgeliefert werden:

```python
# Query 1: Text → Text (Klassisches RAG)
result = multimodal_search(rag, "Wer war Alan Turing?")
mprint(result)

# Query 2: Text → Bild (Textanfrage findet Bilder)
result = multimodal_search(rag, "Zeige mir Bilder von Robotern")
mprint(result)

# Query 3: Bild → Bild (Finde ähnliche Bilder)
similar_images = search_similar_images(rag, "./files/apfel.jpg", k=3)
mprint("## 🖼️ Ähnliche Bilder:")
for img in similar_images:
    mprint(f"  • {img['filename']}: Ähnlichkeit {img['similarity']:.2f}")

# Query 4: Bild → Text (Finde Textinformationen zu einem Bild)
text_result = search_text_by_image(rag, "./files/hedra_cyborg.png", k=3)
mprint(text_result)
```



**Schritt 2: Eigene Queries erstellen**

Entwickle mindestens **3 eigene Queries**, die verschiedene Modalitäten nutzen:

**Beispiele für kreative Queries:**

- **Text→Text**: "Was ist der Unterschied zwischen Robotern und Cyborgs?"
- **Text→Bild**: "Finde alle futuristischen Bilder"
- **Bild→Bild**: Suche ähnliche Bilder zu `a_retro-futuristic_robot_dall_e.jpg`
- **Bild→Text**: Finde Textinformationen, die zum Apfel-Bild passen



**Schritt 3: Bonus-Challenge (Optional)**

Experimentiere mit den Suchparametern:

```python
# Ändere die Anzahl der Ergebnisse (k)
result = multimodal_search(rag, "Roboter", k=5)

# Vergleiche verschiedene Bilder für die Bild→Bild Suche
similar_1 = search_similar_images(rag, "./files/apfel.jpg", k=3)
similar_2 = search_similar_images(rag, "./files/hedra_cyborg.png", k=3)

# Finde heraus, welche Bilder am ähnlichsten zueinander sind
```

# A | Ähnlichkeitsmessung
---


<p><font color='black' size="5">
1. Text-Ähnlichkeit (semantisch)
</font></p>

  Embedding-Modell: OpenAI text-embedding-3-small

  Messmethode:
  - ChromaDB nutzt L2-Distanz (Euklidische Distanz) für Text-Embeddings
  - Werte: 0 = identisch, 2 = maximal entfernt
  - Konvertierung Distanz zu Ähnlichkeit: similarity = max(0, 1 - (score / 2))
  - Threshold: text_threshold: 1.2 (Zeile 51), Mindest-Ähnlichkeit: 0.3

  Verwendung in:
  - search_texts() - Text-Dokumente und Bildbeschreibungen durchsuchen
  - multimodal_search() - Kombinierte Suche

  ---

<p><font color='black' size="5">
2. Bild-Ähnlichkeit (visuell)
</font></p>

  Embedding-Modell: CLIP clip-ViT-B-32

  Messmethode:
  - ChromaDB nutzt Cosine-Distanz für Bild-Embeddings (Zeile 131: "hnsw:space": "cosine")
  - Werte: 0 = identisch, 2 = maximal entfernt
  - Konvertierung zu Ähnlichkeit:
  similarity = round(max(0, 1 - distance), 3)
  - Threshold: image_threshold: 0.8

  Verwendung in:
  - search_images() - Text → Bild Suche über CLIP   
  - search_similar_images() - Bild → Bild Suche  

  ---
  

<p><font color='black' size="5">
3. Cross-Modal-Retrieval (Text ↔ Bild)
</font></p>


  Methode: Indirekte Ähnlichkeit über Bildbeschreibungen

  Ablauf:
  1. Text → Bild: Text-Suche findet Bildbeschreibungen → verknüpfte Bilder werden abgerufen  
  2. Bild → Text: Bild-Suche findet ähnliche Bilder → deren Beschreibungen werden für semantische Textsuche verwendet

  Verknüpfung: Beide Collections sind über text_doc_id ↔ image_doc_id referenziert

  ---
  

<p><font color='black' size="5">
Zusammenfassung der Metriken:
</font></p>

  | Modalität | Embedding-Modell              | Distanzmetrik  | Threshold | Ähnlichkeitsbereich |
  |-----------|-------------------------------|----------------|-----------|---------------------|
  | Text      | OpenAI text-embedding-3-small | L2-Distanz     | 1.2       | 0.3 - 1.0           |
  | Bild      | CLIP ViT-B-32                 | Cosine-Distanz | 0.8       | 0.0 - 1.0           |

  Die Konvertierung 1 - (distance / 2) normalisiert beide Distanzmaße auf einen Ähnlichkeitswert von 0 bis 1, wobei 1 = maximale Ähnlichkeit bedeutet.